test/cbmc/proofs/TCP/prvTCPReturnPacket/TCPReturnPacket_harness.c

```diff
@@ -34,82 +34,136 @@
 #include "FreeRTOS_IP.h"
 #include "FreeRTOS_IP_Private.h"
 #include "FreeRTOS_TCP_IP.h"
+#include "FreeRTOS_TCP_Transmission.h"
 
 #include "../../utility/memory_assignments.c"
 
 #include "cbmc.h"
 
-/* This proof assumes that pxDuplicateNetworkBufferWithDescriptor is implemented correctly. */
-void publicTCPReturnPacket( FreeRTOS_Socket_t * pxSocket,
-                            NetworkBufferDescriptor_t * pxNetworkBuffer,
-                            uint32_t ulLen,
-                            BaseType_t xReleaseAfterSend );
+/*
+ * This function is implemented by a third party.
+ * After looking through a couple of samples in the demos folder, it seems
+ * like the only shared contract is that you want to add the if statement
+ * for releasing the buffer to the end. Apart from that, it is up to the vendor,
+ * how to write this code out to the network.
+ */
+BaseType_t NetworkInterfaceOutputFunction_Stub( struct xNetworkInterface * pxDescriptor,
+                                                NetworkBufferDescriptor_t * const pxNetworkBuffer,
+                                                BaseType_t xReleaseAfterSend )
+{
+    __CPROVER_assert( pxDescriptor != NULL, "The network interface cannot be NULL." );
+    __CPROVER_assert( pxNetworkBuffer != NULL, "The network buffer descriptor cannot be NULL." );
+    __CPROVER_assert( pxNetworkBuffer->pucEthernetBuffer != NULL, "The ethernet buffer cannot be NULL." );
+
+    if( xReleaseAfterSend != pdFALSE )
+    {
+        vReleaseNetworkBufferAndDescriptor( pxNetworkBuffer );
+    }
+
+    return 0;
+}
+
+/* Abstraction of this functions creates an endpoint and assign it to network interface
+ * endpoint, real endpoint doesn't matter in this test. */
+void prvTCPReturn_SetEndPoint( const FreeRTOS_Socket_t * pxSocket,
+                               NetworkBufferDescriptor_t * pxNetworkBuffer,
+                               size_t uxIPHeaderSize )
```

Comment:
`pxNetworkBuffer` is used without NULL check in the actual implementation of the `prvTCPReturn_SetEndPoint()`. It would be better to assert that its never going to be NULL in the proof.